### DISCUSSION TITLE
SPI は LED MATRIX 専用では勿体ないと思いましたが…

### DISCUSSION THREAD

**Discussion by TakSan0** [algyan/XIAOGYAN#42] [General]
今更ではありますが、せっかくアンプとスピーカーがあるのだから音声・音楽も出力やってみようか思って、それにはファイルアクセスしたいしとなって、LED ドライバ (TM1640) のデータシート見ると、この石は Chip ENable 端子の様なものが無いので、常に動いてしまうんですね。
SPI に並列で MicroSD とかを つなごうとしたら、SD にアクセスするたびに LED にちらちらと影響与えてしまうことになると…
 
 Chip Enable があるのなら、SD 等をどうしても使いたい人は IO-Expander を付けて ChipEnable 端子を割り付けてとかできると、
 ちょっと思ったんですが、それも厳しそうですね。
 一応 Static 点灯らしいので、 SPI CLK さえ入れなけいようにすれば分離は出来るのかもしれませんが、 MISO 側の端子も割り当てないといけないですし、ちょっとややこしいし今更感ありますし。
 
 あと、Storage があれば、 SSL 通信用の 証明書関連とかも入れておけますが、無ければハードコーディングしかないとか使いたい人もいるのかな？とも思いました。

 ちょっと思っただけですので、１意見として採用していただかなくもかまいませんが、Storage なかったら、データは Bluetooth 経由か、シリアルや WiFi 経由かしか使えないので、扱いづらいかなと思いました。

**Comment by y-ohe** [comment]
> あと、Storage があれば、 SSL 通信用の 証明書関連とかも入れておけますが、無ければハードコーディングしかないとか使いたい人もいるのかな？とも思いました。
> ちょっと思っただけですので、１意見として採用していただかなくもかまいませんが、Storage なかったら、データは Bluetooth 経由か、シリアルや WiFi 経由かしか使えないので、扱いづらいかなと思いました。

XIAO ESP32C3のArduinoライブラリには、EEPROMライブラリがあるので、これでStorageとして利用できませんか？
このEEPROMライブラリは、内部FLASHを擬似的にEEPROMのようにエミュレーションして使える様にしているみたいなので、完全なEEPROM相当ではないのと、それほど大容量のStorageではないことが制限になりますが、SSL通信用の証明書関連の保管程度なら十分に使えるかと。

**Comment by matsujirushi** [comment]
> それにはファイルアクセスしたいしとなって

(Seeed Studio XIAO ESP32C3なら)SPIFFSが使えませんか？
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/spiffs.html

**Comment by matsujirushi** [comment]
> SPI は LED MATRIX 専用では勿体ないと

もう、フルでXIAOピン割当てしているので、いまの機能を残したままSPIデバイスを追加できるようにするのは無理かなと思います。
（どの機能を選択するのか？という話なので、、、いまある機能と比較して、もっと優先すべきSPIデバイスがあれば変更したいなと思います。）
